pulp_ansible/app/serializers.py

```diff
@@ -361,130 +373,259 @@ class Meta:
 
 class CollectionVersionUploadSerializer(SingleArtifactContentUploadSerializer):
     """
-    A serializer for CollectionVersion Content.
+    A serializer with the logic necessary to upload a CollectionVersion.
+
+    Used in ``.viewsets.CollectionVersionViewSet`` and ``.galaxy.v3.views.CollectionUploadViewSet``
+    to perform the creation and validation of a CollectionVersion and add to repository if
+    necessary. This serializer is meant to be compliant with
+    ``pulpcore.plugin.viewsets.SingleArtifactContentUploadViewSet`` and thus follows these steps on
+    creation:
+
+    1. ``SingleArtifactContentUploadViewSet.create()`` converts uploaded-file (if present) to an
+        artifact and calls ``validate()`` with request in context
+    2. ``pulpcore.plugin.tasks.general_create`` is dispatched with this serializer and its initial
+        validated data and a deferred context determine by the viewset (default an empty context).
+    3. ``general_create`` calls ``validate()`` again with deferred context now.
+        ``deferred_validate()`` is now called and the upload object (if present) is converted to an
+        artifact.
+    4. ``general_create`` calls ``save()`` which will call ``create()``. ``create`` uses the
+        validated data to create and save the CollectionVersion. If repository is specified the
+        CollectionVersion is then added to the repository.
     """
 
-    name = serializers.CharField(help_text=_("The name of the collection."), max_length=64)
+    expected_sha256 = serializers.CharField(
+        help_text=_("An optional sha256 checksum of the uploaded file."),
+        required=False,
+        write_only=True,
+    )
 
-    namespace = serializers.CharField(
-        help_text=_("The namespace of the collection."), max_length=64
+    expected_name = serializers.CharField(
+        help_text=_("The name of the collection."),
+        max_length=64,
+        required=False,
+        write_only=True,
+    )
+
+    expected_namespace = serializers.CharField(
+        help_text=_("The namespace of the collection."),
+        max_length=64,
+        required=False,
+        write_only=True,
     )
 
-    version = serializers.CharField(help_text=_("The version of the collection."), max_length=128)
+    expected_version = serializers.CharField(
+        help_text=_("The version of the collection."),
+        max_length=128,
+        required=False,
+        write_only=True,
+    )
 
     def validate(self, data):
-        """Validate that we have a file or can create one."""
-        if "artifact" in data:
-            raise serializers.ValidationError(_("Only 'file' may be specified."))
+        """Check and set the namespace, name & version."""
+        # This validation is for creating CollectionVersions
+        if self.instance:
+            return data
```

Comment:
Shouldn't we still call `super` first?

Reply:
I'm thinking of removing this check and instead over-riding the `update` method to raise an error if called. `validate` should only be called on `create` or `update` so if `self.instance` is set then `update` will be called and we shouldn't allow it.

Reply:
Hmm. Hard question. From the API we should never allow updates. But we want to allow for a repair command. And it would be super neat if that repair command could just use this serializer to actually _call_ the `process_collection` part again. No?

Reply:
Yes that would be a good use. There would still need to be other fields that would need to be set if you wanted to call `super` in that case (`artifact`, `expected_namespace/name/version/`) else the serializer would fail validation. I'll remove the check and put a TODO comment to revisit this for a future issue (#1106).